File: C++/Binary Search/875_koko-eating-bananas.py
from typing import List
import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        
        while l < r:
            k = l + (r-l) // 2
            res = 0
            for i in piles:
                res += math.ceil(i / k)

            if res > h:
                l = k + 1
            else:
                r = k

        return l
    
# Time Complexity: O(n log m)
# Space Complexity: O(1)

File: solutions/Binary Search/875_koko-eating-bananas.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = *max_element(piles.begin(), piles.end());
        int res;

        while (l <= r) {
            int k = l + (r - l) / 2;

            if (timeTaken(piles, k) <= h) {
                r = k - 1;
                res = k;
            } else {
                l = k + 1;
            }
        }

        return res;
    }

    long long timeTaken(const vector<int>& piles, int s) {
        long long ans = 0;

        for (int p: piles) {
            ans += (p / s) + (p % s > 0);
        }

        return ans;
    }
};